DogScepterCLI/ProjectShell.cs

```diff
@@ -34,238 +70,317 @@ public Command(string[] names, string description, string usage, Func<string[],
-                    }
-                    return Command.CommandResult.None;
-                }),
+                args => AddCommand(console, projectFile, args)),
 
             new Command(new[] { "apply" },
                 "Applies the project to the input data file, resulting in output.",
                 "apply",
-                args =>
-                {
-                    try
-                    {
-                        pf.LoadAll();
-
-                        using FileStream fs = new FileStream(Path.Combine(cfg.OutputDirectory, pf.DataHandle.Filename), FileMode.Create);
-                        using GMDataWriter writer = new GMDataWriter(pf.DataHandle, fs, fs.Name, pf.DataHandle.Length);
-
-                        console.Output.WriteLine("Converting to data...");
-                        pf.ConvertToData();
-                        console.Output.WriteLine("Writing main data file...");
-                        writer.Write();
-                        writer.Flush();
-                        foreach (GMWarning w in writer.Warnings)
-                            console.Output.WriteLine($"[WARN: {w.Level}] {w.Message}"); // todo formatting
-                    }
-                    catch (Exception e)
-                    {
-                        console.Error.WriteLine($"Failed to apply project: {e}");
-                    }
-
-                    if (!ReloadProject(console, ref pf, cfg, verbose))
-                        return Command.CommandResult.Quit;
-
-                    return Command.CommandResult.None;
-                }),
+                _ => ApplyCommand(console, ref projectFile, projectConfig, verbose)),
 
             new Command(new[] { "about" },
                 "Displays information about the open project.",
                 "about",
-                args =>
-                {
-                    console.Output.Write("Data file location: ");
-                    console.Output.WriteLine(cfg.InputFile);
-                    console.Output.Write("Output directory: ");
-                    console.Output.WriteLine(cfg.OutputDirectory);
-                    return Command.CommandResult.None;
-                })
+                _ => AboutCommand(console, projectConfig))
         };
+
         int helpLength = commands.Max(c => c.Usage.Length) + 1;
 
-        console.Error.WriteLine();
-        console.Error.WriteLine("DogScepter project shell");
+        console.Output.WriteLine();
+        console.Output.WriteLine("DogScepter project shell");
 
+        // Enter loop of the shell, until user inputs a command that exits.
         bool running = true;
         while (running)
         {
             console.Output.Write("> ");
-            string command = console.Input.ReadLine();
-            string[] args = command.Split(' ', StringSplitOptions.RemoveEmptyEntries);
+            string input = console.Input.ReadLine();
+            // User likely entered null on purpose, so we let them quit.
```

Comment:
Are you sure about this? Is this null on empty lines or something? If so, that could definitely happen on accident.

Reply:
From MSDN
> If the `Ctrl`+`Z` key combination (followed by `Enter` on Windows) is pressed when the method is reading input from the console, the method returns null. This enables the user to prevent further keyboard input when the [ReadLine](https://docs.microsoft.com/en-us/dotnet/api/system.console.readline?view=net-6.0) method is called in a loop.

Reply:
Alright, everything else looks good to me, then! I'll merge this now.